### ISSUE TITLE
StackOverflow in LogHandler with classloading due to logging in WebAppClassLoader

### ISSUE THREAD

**Issue by oleg-nenashev** [eclipse/jetty.project#1563]
Hello,

In Jenkins project we have recently upgraded from Jetty 9.2 to 9.4. In some conditions we hit StackOverflow in Jetty WebAppClassLoader when Jetty Debug logging is enabled. Issue - [JENKINS-44330](https://issues.jenkins-ci.org/browse/JENKINS-44330). It happens, because one of our LogAppenders does classloading:

```
at org.eclipse.jetty.util.log.JavaUtilLog.log(JavaUtilLog.java:167)
at org.eclipse.jetty.util.log.JavaUtilLog.debug(JavaUtilLog.java:227)
at org.eclipse.jetty.webapp.WebAppContext.isSystemResource(WebAppContext.java:857)
at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:548)
at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
at hudson.logging.LogRecorder.orderedTargets(LogRecorder.java:82)
at hudson.logging.LogRecorder$2.publish(LogRecorder.java:108)
at hudson.logging.WeakLogHandler.publish(WeakLogHandler.java:53)
at java.util.logging.Logger.log(Logger.java:738)
at org.eclipse.jetty.util.log.JavaUtilLog.log(JavaUtilLog.java:167)
at org.eclipse.jetty.util.log.JavaUtilLog.debug(JavaUtilLog.java:227)
at org.eclipse.jetty.webapp.WebAppContext.isSystemResource(WebAppContext.java:857)
at org.eclipse.jetty.webapp.WebAppClassLoader.loadClass(WebAppClassLoader.java:548)
at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
at hudson.logging.LogRecorder.orderedTargets(LogRecorder.java:82)
at hudson.logging.LogRecorder$2.publish(LogRecorder.java:108)
at hudson.logging.WeakLogHandler.publish(WeakLogHandler.java:53)
```

I think we should fix this regression on our side. OTOH I would like to record this issue here, because it may hit other projects. Jetty project contributors may want to apply a patch as well.

CC @olamy @damphyr

**Comment by oleg-nenashev** [comment]
So it will happen if and only if the Debug logging for Jetty is enabled: https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java#L857

It has been added by this commit: https://github.com/eclipse/jetty.project/commit/afb750515e3f5bd2f3ebb7a84b5dc34f7cdcef6a . Will raise an issue in the Jetty bugtracker

**Comment by gregw** [comment]
Tough call how to fix.  Not sure i like any alternative: leave it to log
implementor to avoid;  remove logging from classloader; detect recursion in
classloader; detect recursion in logging.

**Comment by joakime** [comment]
@gregw perhaps we should have a recursion check, this isn't the first StackOverflow from this new classpath pattern logic. See Issue #1448

**Comment by oleg-nenashev** [comment]
We fixed the issue on our side, but recursion prevention would be definitely useful. In the case of Debug logging it should also prevent issues with performance though #1448 won't be addressed then

**Comment by gregw** [comment]
So if we detect recursion in classloading, then what do we do?  Guess it is because of logging and don't log during the load?   Seams a bit contrived?

It is much harder for us, perhaps impossible to put recursion detection into logging, because we don't implement all the log handlers and it is a big tax to pay on log handlers that can never be recursive.

Still not convinced we can do anything reasonable.

**Comment by joakime** [comment]
Would it be possible to put the recursion check in `WebAppClassLoader.loadClass()`?

**Comment by gregw** [comment]
Not really:   
* `WACL` is as asked to load a class
* it does some logging
* the log handler needs to load a class
* recursion is detected

Then what? return null? throw exception? if so who catches it? How is logging ever instantiated because it can also happen:

* something does some logging
* the log handler needs to load a class
*  `WACL` is as asked to load the class
* it does some logging
* the log handler needs to load a class
* recursion is detected

Basically logging can't be implemented in terms of anything that actually does logging!

Perhaps that means the ClassLoader should not do any logging?
